cirq-google/cirq_google/circuit_serializer.py

```diff
@@ -0,0 +1,657 @@
+from cirq_google.ops import PhysicalZTag
+from cirq_google.ops.calibration_tag import CalibrationTag
+
+
+class CircuitSerializer:
+    """A class for serializing and deserializing programs and operations.
+
+    This class is for serializing cirq_google.api.v2. protos using one
+    message type per gate type.  It serializes qubits by adding a field
+    into the constants table.  Usage is by passing a `cirq.Circuit`
+    to the `serialize()` method of the class, which will produce a
+    `Program` proto.  Likewise, the `deserialize` method will produce
+    a `cirq.Circuit` object from a `Program` proto.
+
+    This class is more performant than the previous `SerializableGateSet`
+    at the cost of some extendability.
+    """
+
+    def __init__(
+        self,
+        gate_set_name: str,
+    ):
+        """Construct the gate set.
+
+        Args:
+            gate_set_name: The name used to identify the gate set.
+            serializers: The OpSerializers to use for serialization.
+                Multiple serializers for a given gate type are allowed and
+                will be checked for a given type in the order specified here.
+                This allows for a given gate type to be serialized into
+                different serialized form depending on the parameters of the
+                gate.
+            deserializers: The OpDeserializers to convert serialized
+                forms of gates or circuits into Operations.
+        """
+        self.gate_set_name = gate_set_name
+
+    def serialize(
+        self,
+        program: cirq.Circuit,
+        msg: Optional[v2.program_pb2.Program] = None,
+        *,
+        arg_function_language: Optional[str] = None,
+        use_constants: bool = True,
+    ) -> v2.program_pb2.Program:
+        """Serialize a Circuit to cirq_google.api.v2.Program proto.
+
+        Args:
+            program: The Circuit to serialize.
+            msg: An optional proto object to populate with the serialization
+                results.
+            arg_function_language: The `arg_function_language` field from
+                `Program.Language`.
+            use_constants: Whether to use constants in serialization. This is
+                required to be True for serializing CircuitOperations.
+        """
+        if msg is None:
+            msg = v2.program_pb2.Program()
+        msg.language.gate_set = self.gate_set_name
+        if isinstance(program, cirq.Circuit):
+            constants: Optional[List[v2.program_pb2.Constant]] = [] if use_constants else None
+            raw_constants: Optional[Dict[Any, int]] = {} if use_constants else None
+            self._serialize_circuit(
+                program,
+                msg.circuit,
+                arg_function_language=arg_function_language,
+                constants=constants,
+                raw_constants=raw_constants,
+            )
+            if constants is not None:
+                msg.constants.extend(constants)
```

Comment:
Could we instead pass `msg.constants` into `_serialize_circuit` so that the function can add constants to it directly? Perhaps this could be combined with `raw_constants` into a helper object that deals with the details of storing constants, since `constants` and `raw_constants` must both be present or not.

Reply:
Passed msg.constants into the file now.  Decided not to make a helper object given that raw_constants is the helper object now, and, with all the refactoring, everything that uses these is now private.